python/lsst/daf/butler/instrument.py

```diff
@@ -47,6 +49,21 @@ class Instrument(metaclass=ABCMeta):
     a no-argument callable that can be used to construct a Python instance.
     """
 
+    configPaths = []
+    """Paths to config files to read for specific Tasks.
+
+    The paths in this list should contain files of the form `task.py`, for
+    each of the Tasks that requires special configuration.
+    """
+
+    filterDefinitions = None
+    """`~lsst.obs.base.FilterDefinitions`: The definitions of the filters used
+    by this instruments.
+    """
+
+    def __init__(self, *args, **kwargs):
+        self.filterDefinitions.defineFilters()
```

Comment:
I think I'm going to need some advice on how to manage this: I suspect this is the cause of my current ci_hsc failures (the afw `Filters` not being defined at the right time).

Reply:
I think it would work to put this in the constructor for the HSC raw data formatter, behind some kind of guard that checks that it's only ever done once.  Or at least that's a place where it should go; there may be other places (also behind the same call-once guard).

Now that I think about it, it may make sense to put that guard in `FilterDefinitionCollection.defineFilters`, and make it so the behavior is that:
 - calling `defineFilters` multiple times on the same `FilterDefinitionCollection` is a silent no-op;
 - calling `defineFilters` on different instances of `FilterDefinitionCollection` is an error.  That's unfortunate, but I don't think we can expect trying to put multiple instruments' filters into the afw singleton to lead to anything but pain.

Reply:
Why not put that `defineFilters` call into `fitsRawFormatterBase.__init__`? Otherwise, it would have to go into every child class's `__int__`. I think we need to have some guarantee that the filters are defined *somewhere*, and enforcing it in a parent class seems the most reasonable to me.

Reply:
This is all going to get very difficult once we start having gen 3 repositories with multiple instruments in them.

Reply:
> Why not put that defineFilters call into `fitsRawFormatterBase.__init__`?

:+1:

I still think that'll be necessary but not sufficient, as it doesn't solve the problem with having filters defined before non-`raw` `Exposures` are read.  I don't know how to deal with that (or how we've avoided dealing with it) so far.  Maybe the `Exposure` read code is more robust against undefined filters than I'm guessing.

> This is all going to get very difficult once we start having gen 3 repositories with multiple instruments in them.

Not just difficult.  I consider multi-instrument repositories effectively blocked on the removal of the `Filter` singleton in afw.